flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -820,16 +796,13 @@ private void doRun() {
                             || current == ExecutionState.DEPLOYING) {
                         if (t instanceof CancelTaskException) {
                             if (transitionState(current, ExecutionState.CANCELED)) {
-                                cancelInvokable(invokable);
                                 break;
                             }
                         } else {
                             if (transitionState(current, ExecutionState.FAILED, t)) {
                                 // proper failure of the task. record the exception as the root
                                 // cause
                                 failureCause = t;
-                                cancelInvokable(invokable);
-
                                 break;
```

Comment:
Can we end up here? Isn't it the case that if there is an exception we will return from `restoreAndInvoke` in either `CANCELLING` or `FAILED` state?